### ISSUE TITLE
Cannot compile shades error on `ErrorCannotLensIntoOptionalKey`

### ISSUE THREAD

**Issue by wcastand** [jamesmcnamara/shades#37]
I get this error when i try to use `tsc`:

```bash

../node_modules/shades/types/utils.ts:28:41 - error TS6205: All type parameters are unused

28 interface ErrorCannotLensIntoOptionalKey<T, K> {
                                           ~~~~~~


Found 1 error.

error Command failed with exit code 2.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

i think it should be quite easy to fix, i checked and the params are not use and `ErrorCannotLensIntoOptionalKey` is only used here. so if you're happy with me deleting the <T, K> on this then i can make the PR.

**Comment by jamesmcnamara** [comment]
Hmm so this is very intentional. Say you have an object with a potentially undefined key like:
```ts
interface Post {
  text: string;
  comments?: Comment[];
}

interface Comment {
  text: string;
  creator: Person;
}
```
And you tried to lens through the `comments` field without checking for undefined:
```
get('comments', all, 'creator')(myPost)
```
This should be a type error because it could crash at run time. But by default TS gives a very complex type error that is very difficult to understand. So I made this type `ErrorCannotLensIntoOptionalKey<T, K>` that is instantiated with the current value type T and the path for K. So then the type error you would get would displayed as:
```
Some TS Error on:
interface ErrorCannotLensIntoOptionalKey<Post, 'comments'> {
  error: 'You have tried to lens through an optional key. Consider using `fill` to provide defaults to your object';
}
```

What version of TS are you using? Are you using TS/ESLint? Maybe we can find a modification that works more generally.

**Comment by jamesmcnamara** [comment]
Hmmm so this is intentional for editor usability. When you try to get or mod past a field that is undefined you will get a very inscrutable TS error.

**Comment by wcastand** [comment]
Thank for the explanation, it makes a lot of sense. :)
i started to get the error when i switch to typescript 8 (`"typescript": "3.8.2"`)
So maybe they have new rules in the last version against unused generics.
For now, I've patched it in our codebase with `patch-package` so i can compile my project.


i'm not using a linter, the error only appears at compile time with `tsc`

this is my `tsconfig.json`:
```json
{
  "compilerOptions": {
    "declaration": true,
    "declarationDir": "./dist",
    "strict": true,
    "allowSyntheticDefaultImports": true,
    "sourceMap": true,
    "lib": ["dom", "dom.iterable", "esnext"],
    "noUnusedParameters": true,
    "strictNullChecks": true,
    "skipLibCheck": true,
    "moduleResolution": "node",
    "noImplicitAny": true,
    "outDir": "./dist",
    "target": "es5",
    "module": "CommonJS",
    "esModuleInterop": true,
    "typeRoots": ["node_modules/@types"],
    "jsx": "react"
  },
  "include": ["src/**/*"],
  "exclude": ["node_modules", "build", "dist", "scripts"]
}
```

Let me know if i can help or if you need more infos :)

**Comment by jamesmcnamara** [comment]
Yeah 3.8 is causing a lot of problems. Maybe I can do something with phantom parameters. 

If you wanted to try something, you could take that PR and add:
```ts
 interface ErrorCannotLensIntoOptionalKey<T, K> {
  error: ...,
  _val: T,
  _key: K
}
```
and see if that helps. That type is never constructed so we might not have to do anything else.

**Comment by wcastand** [comment]
Just pushed and i tested on the repo where i was getting the error and it seems to work, no error anymore.

**Comment by wcastand** [comment]
Thanks for merging the fix, do we wait for other issues to be fixed before publishing?

**Comment by icylace** [comment]
Please publish !

**Comment by jamesmcnamara** [comment]
Resolved in shades@2.2
